helix-core/src/main/java/org/apache/helix/tools/ClusterVerifiers/BestPossibleExternalViewVerifier.java

```diff
@@ -274,8 +276,24 @@ protected synchronized boolean verifyState() {
 
       // Filter resources if requested
       if (_resources != null && !_resources.isEmpty()) {
-        idealStates.keySet().retainAll(_resources);
-        extViews.keySet().retainAll(_resources);
+        Set<String> wagedResource = _resources.stream().filter(
+                resourceEntry -> WagedValidationUtil.isWagedEnabled(idealStates.get(resourceEntry)))
+            .collect(Collectors.toSet());
+        if (wagedResource.isEmpty()) {
+          // If no waged-enabled resources are found, retain only the provided resources
+          idealStates.keySet().retainAll(_resources);
+          extViews.keySet().retainAll(_resources);
```

Comment:
These two lines are similar to following. Simplify the entire if-else block to be:

Set<String> resourcesToRetain = new HashSet<>(_resources);
if (!wagedResource.isEmpty()) {
 resourcesToRetain.addAll( idealStates.keySet().stream().filter(
                  resourceEntry -> WagedValidationUtil.isWagedEnabled(idealStates.get(resourceEntry)))
              .collect(Collectors.toSet()));
}

idealStates.keySet().retainAll(resourcesToRetain);
extViews.keySet().retainAll(resourcesToRetain);